pkg/azurefile/controllerserver.go

```diff
@@ -1088,6 +1089,90 @@ func (d *Driver) ListSnapshots(_ context.Context, _ *csi.ListSnapshotsRequest) (
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
+// restoreSnapshot restores from a snapshot
+func (d *Driver) restoreSnapshot(ctx context.Context, dstAccountName string, req *csi.CreateVolumeRequest, accountKey string, shareOptions *fileclient.ShareOptions, storageEndpointSuffix string) error {
+	if shareOptions.Protocol == storage.EnabledProtocolsNFS {
+		return fmt.Errorf("protocol nfs is not supported for snapshot restore")
+	}
+	var sourceSnapshotID string
+	if req.GetVolumeContentSource() != nil && req.GetVolumeContentSource().GetSnapshot() != nil {
+		sourceSnapshotID = req.GetVolumeContentSource().GetSnapshot().GetSnapshotId()
+	}
+	resourceGroupName, srcAccountName, srcFileShareName, _, _, _, err := GetFileShareInfo(sourceSnapshotID) //nolint:dogsled
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	snapshot, err := getSnapshot(sourceSnapshotID)
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	dstFileShareName := shareOptions.Name
+	if srcFileShareName == "" || dstFileShareName == "" {
+		return fmt.Errorf("srcFileShareName(%s) or dstFileShareName(%s) is empty", srcFileShareName, dstFileShareName)
+	}
+
+	srcAccountKey := accountKey
+	if srcAccountName != dstAccountName {
+		srcAccountKey, err = d.cloud.GetStorageAccesskey(ctx, d.cloud.SubscriptionID, srcAccountName, resourceGroupName, true)
+		if err != nil {
+			return fmt.Errorf("get src accountkey get err: %v", err)
+		}
+	}
+
+	klog.V(2).Infof("generate sas token for resource group(%s) src account(%s)", resourceGroupName, srcAccountName)
+	srcAccountSasToken, err := generateSASToken(srcAccountName, srcAccountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes)
```

Comment:
should use current new way to get azcopy auth, e.g. always use managed identity by default if secrets not provided